WIP:new Reverse Proxy: Traefik

Kukks opened the pull request:
This PR adds support to run BTCPayServer behind traefik.
You either set the reverse proxy env var to "traefik" or else specify an additional fragment "traefik-labels" which tags the btcpayserver server and allows it to be discovered by an already running traefik container.

Just need to test. :)

ch4ot1c commented on Production/traefilk.toml:
```diff
@@ -0,0 +1,35 @@
+defaultEntryPoints = ["https","http"]
+
```
Filename error: `traefilk.toml` -> `traefik.toml`

ch4ot1c commented on Production/traefilk.toml:
```diff
@@ -0,0 +1,35 @@
+defaultEntryPoints = ["https","http"]
+
+logLevel="ERROR"
+
```
Nit: spacing consistency (`x = "val"`)

ch4ot1c commented:
utACK. Cool!

NicolasDorier commented on build.sh:
```diff
@@ -22,3 +22,11 @@ docker run -v "$(pwd)/Generated:/app/Generated" \
+
+
+
+
```
empty lines

NicolasDorier commented on docker-compose-generator/src/docker-compose-generator.csproj:
```diff
@@ -10,4 +10,33 @@
 	  <PackageReference Include="YamlDotNet" Version="4.3.1" />
   </ItemGroup>
 
+  <ItemGroup>
```
why adding this to csproj?

NicolasDorier commented:
need proper doc

Kukks commented:
yes @NicolasDorier  I was just going through it first and seeing what was missing and needed.
* New pregen folder `Production-Traefik`
* Nginx settings on btcpayserver not emitted on no `reverse proxy mode`(moved to nginx fragment)
* Split postgres and nbxplorer to their own fragments
* added support to use our own image of the generator through env ` $BTCPAYGEN_DOCKER_IMAGE` for testing.
Will add docs today

Kukks commented:
Just tested this out @ https://btcpay.kukks.org and it works!

Kukks commented:
@NicolasDorier  this one is ready too :)

NicolasDorier commented:
Remove the pre generated docker-compose. Just add fragments. I removed them yestersday, it was legacy stuff.

Also remove `Production-` from your folder name, this was also legacy stuff

NicolasDorier commented:
Good idea to break down the docker-compose.

Kukks commented:
Done. Just  a note, this one include the work done in https://github.com/btcpayserver/btcpayserver-docker/pull/42 as it was helpful to test

NicolasDorier commented:
@Kukks shit I saw too late that you documented here, and documented it on my own. So now there is conflict :(

NicolasDorier commented:
You can remove the part in `How can I add an altcoin to BTCPayServer?`. Also I rewrote a bit your code. So rebase your PR on the master.

Kukks commented:
too many branches :D should be up to date now

NicolasDorier commented on build.ps1:
```diff
@@ -1,4 +1,5 @@
 # This script will run docker-compose-generator in a container to generate the yml files
+if (-not ($BTCPAYGEN_DOCKER_IMAGE)) { $BTCPAYGEN_DOCKER_IMAGE = "btcpayserver/docker-compose-generator" }
```
Remove this line

NicolasDorier commented on btcpay-setup.sh:
```diff
@@ -71,6 +71,7 @@ fi
 
 ######### Migration: old pregen environment to new environment ############
 if [ ! -z $BTCPAY_DOCKER_COMPOSE ] && [ ! -z $DOWNLOAD_ROOT ] && [ -z $BTCPAYGEN_OLD_PREGEN ]; then 
+    if [[ $(dirname $BTCPAY_DOCKER_COMPOSE) == *Production-NoReverseProxy ]]; then
```
remove this line

NicolasDorier commented on btcpay-setup.sh:
```diff
@@ -51,16 +51,16 @@ Environment variables:
     LETSENCRYPT_EMAIL: A mail will be sent to this address if certificate expires and fail to renew automatically (eg. me@example.com)
     NBITCOIN_NETWORK: The type of network to use (eg. mainnet, testnet or regtest. Default: mainnet)
-    LIGHTNING_ALIAS: An alias for your lightning network node if used
+    LIGHTNING_ALIAS: An alias for your lightning network node if used    
```
space

NicolasDorier commented on Traefik/README.md:
```diff
@@ -0,0 +1,73 @@
+
+If `BTCPAY_HOST` is `btcpay.example.com` and `BTCPAY_ROOTPATH` is `/btcpay`, then you can access the site via `https://btcpay.example.com/btcpay`
+
+Use `docker-compose.btc-ltc.yml` for bitcoin and litecoin support, or `docker-compose.btc.yml` for only bitcoin.
```
does not exist anymore

NicolasDorier commented on build.sh:
```diff
@@ -29,3 +29,9 @@ docker run -v "$(pwd)/Generated:/app/Generated" \
+
+if [ "$BTCPAYGEN_REVERSEPROXY" == "traefik" ]; then
+    cp Traefik/traefik.toml Generated/traefik.toml
+    :> Generated/acme.json
```
what is `:>`? a smiley?

Kukks replied:
That one ensures that there is an empty file at `Generated/acme.json`

NicolasDorier commented on Traefik/README.md:
```diff
@@ -0,0 +1,71 @@
+# About those docker-compose
+
+All `docker-compose` files in [Production](Production), [Production-NoReverseProxy](Production-NoReverseProxy) and [Production-Traefik](Production-Traefik) are generated by running the [build-pregen.sh](build-pregen.sh) (or [build-pregen.ps1](build-pregen.ps1)) script from the fragments located in [docker-compose-generator/docker-fragments](docker-compose-generator/docker-fragments).
```
Remove this part, build-pregen is dead.

NicolasDorier commented on Traefik/README.md:
```diff
@@ -0,0 +1,71 @@
+
+---
+
+The `docker-compose` can be used for production purpose.
```
Remove this part, build-pregen is dead.

NicolasDorier commented:
You can delete most of `Traefik/README.md` which is just copy pasted stuff.

Just explain if there is specific environment variable for traefik, and what traefik bring versus nginx.

iBobik commented on docker-compose-generator/docker-fragments/traefik-labels.yml:
```diff
@@ -0,0 +1,12 @@
+      - "traefik.frontend.rule=Host:${BTCPAY_HOST}"
+      - "traefik.port.rule=49392"
+      - "traefik.acme.domains=${BTCPAY_HOST},www.${BTCPAY_HOST}"
+      - "traefik.acme.email=${LETSENCRYPT_EMAIL}"
```
Why all this settings? It should be enough to just specify this:
- "traefik.enable=true"
- "traefik.frontend.rule=Host:${BTCPAY_HOST}"
As I know others are optional or can be set in traefik.toml.

Kukks replied:
This is from 2018 lol. We'll most likely remove traefik support since it's not feasible to maintain 2 separate proxy options for minimal benefits

iBobik commented:
Let's discuss Trafic there, please: https://github.com/btcpayserver/btcpayserver-docker/issues/570

Honza Pobořil


> 20. 12. 2021 v 8:20, Andrew Camilleri ***@***.***>:
> 
> ﻿
> @Kukks commented on this pull request.
> 
> In docker-compose-generator/docker-fragments/traefik-labels.yml:
> 
> > @@ -0,0 +1,12 @@
> +version: "3"
> +
> +services:
> +  btcpayserver:
> +    labels:
> +      - "traefik.backend=btcpayserver"
> +      - "traefik.backend.loadbalancer.sticky=true"
> +      - "traefik.enable=true"
> +      - "traefik.frontend.rule=Host:${BTCPAY_HOST}"
> +      - "traefik.port.rule=49392"
> +      - "traefik.acme.domains=${BTCPAY_HOST},www.${BTCPAY_HOST}"
> +      - "traefik.acme.email=${LETSENCRYPT_EMAIL}"
> This is from 2018 lol. We'll most likely remove traefik support since it's not feasible to maintain 2 separate proxy options for minimal benefits
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> Triage notifications on the go with GitHub Mobile for iOS or Android. 
> You are receiving this because you commented.
